BJS2-77992. Добавлено создание цели

FROLOV-KI opened the pull request:
1) Добавлен GoalService
2) Добавлен GoalController
3) Добавлен GoalMapper
4) Добавлены кастомные исключения для сущностей goal и skill 5) Добавлен SkillService
6) Добавлен GoalExceptionHandler

FROLOV-KI committed:
BJS2-77992. Добавлено создание цели

1) Добавлен GoalService
2) Добавлен GoalController
3) Добавлен GoalMapper
4) Добавлены кастомные исключения для сущностей goal и skill
5) Добавлен SkillService
6) Добавлен GoalExceptionHandler

FROLOV-KI committed:
BJS2-77992. Обновлено создание цели

1) Вызов маппера перенесен в контроллер
2) Добавлены классы валидаторы

psvm-dev commented on src/main/java/school/faang/user_service/entity/goal/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,9 @@
+import school.faang.user_service.entity.goal.Goal;
+import school.faang.user_service.entity.goal.GoalDto;
+
+public interface GoalMapper {
```
Можно было бы решить через mapstruct

psvm-dev commented on src/main/java/school/faang/user_service/exception/goal/MaxActiveGoalPerUserException.java:
```diff
@@ -0,0 +1,8 @@
+public class MaxActiveGoalPerUserException extends RuntimeException {
+
+    public MaxActiveGoalPerUserException(Long userId, int goalLimit) {
+        super("User with id %s reach max goals active limit of %d".formatted(userId, goalLimit));
```
лучше сделать
(String message) {super(message);}
и уже при пробросе исключения в конструкторе писать сообщение.

psvm-dev commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -23,6 +24,13 @@ INSERT INTO goal (title, description, parent_goal_id, status, created_at, update
+            INSERT INTO user_goal (user_Id, goal_Id, created_at, updated_at)
+            VALUES (:userId, :goalId, NOW(), NOW())
+            """)
+    void assignGoalToUser(long userId, long goalId);
```
я бы сделала UserGoalRepository extends JpaRepository и использовала его метод save()

20cognosce replied:
Да уж, согласен с оратором выше, не надо всё на нативках писать

FROLOV-KI committed:
BJS2-77992. Добавлено обновление цели

1) Удален GoalExceptionHandler.java
2) Добавлен метод удаление связи цель - скилл

FROLOV-KI committed:
BJS2-77992. Добавлено удаление цели

tanufa25 commented on src/main/java/school/faang/user_service/validator/goal/SkillValidator.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.Collectors;
+
+@Component
+@RequiredArgsConstructor
```
А зачем @RequaredArgsConstructor, если нет никаких полей?
Тут и обычный дефолтный конструктор сработает, который не надо никак прописывать.
Но можешь не исправлять, мне для валидации нужен репозиторий, я его добавлю, а вот с ним уже нужен @RequaredArgsConstructor

FROLOV-KI replied:
Были, я их удалил, а аннотацию забыл

FROLOV-KI committed:
BJS2-77992. Удаление цели обновлено

tanufa25 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+    private final CreateGoalMapperImpl createGoalMapper;
+    private final UpdateGoalMapperImpl updateGoalMapper;
+
+    @PostMapping("/create")
```
PostMapping самодостаточный для create, его лучше без параметров оставить, и по этой url /api/v1/goal делать create

tanufa25 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+                .body(createGoalMapper.goalToDto(createdGoal));
+    }
+
+    @PutMapping("/update/{goalId}")
```
Тоже /update не нужен в пути

tanufa25 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+        return ResponseEntity.ok(updateGoalMapper.goalToDto(createdGoal));
+    }
+
+    @DeleteMapping("/delete/{goalId}")
```
И /delete не нужен в пути

FROLOV-KI committed:
BJS2-77992. Комментарии к ПР

FROLOV-KI committed:
BJS2-77992. Обновил маппер на использование MapStruct

FROLOV-KI committed:
BJS2-77992. Добавил получения списка целей и списка подцелей с фильтрами

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,83 @@
+        return ResponseEntity.ok(goalMapper.toGoalDto(goal));
+    }
+
+    @GetMapping
+    public ResponseEntity<List<GoalDto>> getGoals(@RequestBody GoalFilterDto goalFilterDto) {
```
Post, из-за того что тело передаешь

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,83 @@
+        return ResponseEntity.ok(goalMapper.toGoalDtoList(filteredGoals));
+    }
+
+    @GetMapping("/{parentId}/subGoals")
+    public ResponseEntity<List<GoalDto>> getSubGoals(@PathVariable long parentId, @RequestBody GoalFilterDto goalFilterDto) {
```
тоже post

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,83 @@
+                goalDto.parentId()
+        );
+
+        URI location = ServletUriComponentsBuilder
+                .fromCurrentRequest()
+                .path("/{id}")
+                .buildAndExpand(createdGoal.getId())
+                .toUri();
```
можно было бы это не делать

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,83 @@
+                .body(goalMapper.toGoalDto(createdGoal));
+    }
+
+    @PutMapping("/{goalId}")
```
Patch

20cognosce commented on src/main/java/school/faang/user_service/entity/goal/dto/GoalDto.java:
```diff
@@ -0,0 +1,15 @@
+        Long id,
+        String description,
+        Long parentId,
+        @NotBlank(message = "Empty goal title not allowed!") String title,
```
зачем ты добавил валидацию на DTO, которую сам же возвращаешь
на уровне БД должен быть этот констрейнт

20cognosce commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<GoalDto> createGoal(@RequestBody @Valid GoalDto goalDto) {
```
Я говорил кстати сделать раздельные ДТОшки

20cognosce commented on src/main/java/school/faang/user_service/entity/goal/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,46 @@
+    @Mapping(target = "users", ignore = true)
+    Goal toGoal(GoalDto goalDto);
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "parent", ignore = true)
+    @Mapping(target = "skillsToAchieve", ignore = true)
+    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+    void update(@MappingTarget Goal goal, Goal goalData);
```
апдейт делать руками в сервисе, пожалуйста, это часть бизнес-логики

20cognosce commented on src/main/java/school/faang/user_service/filter/goal/GoalFilterByStatus.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public Stream<Goal> apply(Stream<Goal> filteredData, GoalFilterDto filterDto) {
+        return filteredData.filter(goal -> goal.getStatus().equals(filterDto.status()));
```
goal.getStatus() == null -> NPE

20cognosce replied:
сто раз просил использовать Objects.equals

20cognosce commented on src/main/java/school/faang/user_service/filter/goal/GoalFilterByTitle.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public Stream<Goal> apply(Stream<Goal> filteredData, GoalFilterDto filterDto) {
+        return filteredData.filter(goal -> goal.getTitle()
+                .toLowerCase()
```
.

20cognosce commented on src/main/java/school/faang/user_service/filter/goal/GoalFilterBySkills.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public Stream<Goal> apply(Stream<Goal> filteredData, GoalFilterDto filterDto) {
+        return filteredData.filter(goal ->
+                goal.getSkillsToAchieve()
+                        .stream()
```
.

20cognosce commented on src/main/java/school/faang/user_service/filter/goal/GoalFilterByUsers.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public Stream<Goal> apply(Stream<Goal> filteredData, GoalFilterDto filterDto) {
+        return filteredData.filter(goal ->
+                goal.getUsers()
+                        .stream()
```
.

20cognosce commented on src/main/java/school/faang/user_service/repository/SkillRepository.java:
```diff
@@ -40,6 +40,14 @@ public interface SkillRepository extends JpaRepository<Skill, Long> {
     @Modifying
     void assignSkillToUser(long skillId, long userId);
 
+    @Query(nativeQuery = true, value = "INSERT INTO goal_skill (goal_id, skill_id) VALUES (:goalId, :skillId)")
```
многострочные запросы пиши в текстовых блоках """
каждая команда с новой строки, чтобы красивое форматирование получилось

20cognosce replied:
если это у тебя связная таблица, под нее можно отдельную сущность создать для удобства, вместо нативок

20cognosce commented on src/main/java/school/faang/user_service/repository/SkillRepository.java:
```diff
@@ -40,6 +40,14 @@ public interface SkillRepository extends JpaRepository<Skill, Long> {
+    @Modifying
+    void assignSkillToGoal(long goalId, long skillId);
+
+    @Query(value = "DELETE FROM goal_skill WHERE goal_id = :goalId", nativeQuery = true)
```
для таких простых запросов можно просто по неймингу метода сделать запрос
`deleteGoalSkillByGoalId(Long goalId)`

20cognosce replied:
спринг сам спарсит

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,46 @@
+    private final SkillValidator skillValidator;
+
+    public void assignSkillToGoal(long goalId, List<Long> skillsId) {
+        List<Long> absentSkillsId = skillsId.stream()
+                .filter(this::isSkillNotExists)
+                .toList();
```
почему это наполовину в сервисе, а другая половина в валидаторе, ты уж определись, либо все проверки там, либо все тут

20cognosce commented on src/main/java/school/faang/user_service/validator/goal/SkillValidator.java:
```diff
@@ -0,0 +1,20 @@
+@Component
+public class SkillValidator {
+
+    public void validateExistingSkills(List<Long> absentSkillsId) {
+        if (!absentSkillsId.isEmpty()) {
+            String notExistingSkillsId = absentSkillsId.stream()
+                    .map(String::valueOf)
+                    .collect(Collectors.joining(", "));
+            throw new AddedSkillNotExistException(notExistingSkillsId);
+        }
```
сейчас тут как таковой проверки нет, просто ошибка бросается

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,46 @@
+    private final SkillRepository skillRepository;
+    private final SkillValidator skillValidator;
+
+    public void assignSkillToGoal(long goalId, List<Long> skillsId) {
```
@Transactional spring

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,46 @@
+        skillsId.forEach(skillId -> skillRepository.assignSkillToGoal(goalId, skillId));
+    }
+
+    public void updateSkillForGoal(long goalId, List<Long> newSkillsId) {
```
replace тогда уж, и тоже транзакция

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,46 @@
+        assignSkillToGoal(goalId, newSkillsId);
+    }
+
+    public void assignSkillsToUser(long userId, List<Skill> skills) {
```
транзакция

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,46 @@
+                .forEach(skill -> skillRepository.assignSkillToUser(skill.getId(), userId));
+    }
+
+    public boolean isSkillNotExists(long skillId) {
+        return !skillRepository.existsById(skillId);
+    }
+
+    public List<Skill> findSkillsByGoalId(long goalId) {
+        return skillRepository.findSkillsByGoalId(goalId);
```
Transactional(readOnly = true)

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+
+    private final UserContext userContext;
+
+    private final GoalMapper goalMapper;
```
почему сервис имеет на него зависимость

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+            throw new GoalNotExistException(parentId);
+        }
+
+        Goal newGoal = goalRepository.create(newGoalData.getTitle(), newGoalData.getDescription(), parentId);
```
save
и сущность передавай созданную

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+                        .toList()
+        );
+
+        goalMapper.update(dbGoal, newGoalData);
```
сетишь новые поля и тоже save

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+                .stream()
+                .map(Skill::getId)
+                .toList()
+                .equals(skillsId);
```
эм, а если порядок другой выйдет? это баг будет уже
containsAll

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+                .toList()
+                .equals(skillsId);
+
+        if (skillsChanged) {
+            skillService.updateSkillForGoal(goalId, skillsId);
+            dbGoal.setSkillsToAchieve(skillService.findSkillsByGoalId(goalId));
+        }
```
и это не особо понятные действия
вот был скилл1, скилл2
пришел апдейт на скилл2, скилл3
в результате только скилл2 и скилл3 должны быть назначены, получается, если список не пустой, он всегда назначается и без доп проверок можно обойтись

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+                .equals(skillsId);
+
+        if (skillsChanged) {
+            skillService.updateSkillForGoal(goalId, skillsId);
+            dbGoal.setSkillsToAchieve(skillService.findSkillsByGoalId(goalId));
```
в двух местах одно и то же делать не надо, со стороны цели только делай set, в save как связные сущности должно все подвязаться.

Либо только в репозитории связной таблицы напрямую один запрос выполняй

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+            dbGoal.setSkillsToAchieve(skillService.findSkillsByGoalId(goalId));
+        }
+
+        if (dbGoal.getStatus().equals(GoalStatus.COMPLETED)) {
```
енамы через == проверяются

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+        Goal goal = getGoalById(goalId);
+
+        boolean userNotOwner = goalRepository.findGoalsByUserId(userContext.getUserId())
+                .noneMatch(userGoal -> userGoal.getId().equals(goalId));
```
NPE, невозможно, но все равно, привычку внедряй делать null-safe код

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+        boolean userNotOwner = goalRepository.findGoalsByUserId(userContext.getUserId())
+                .noneMatch(userGoal -> userGoal.getId().equals(goalId));
+
+        if(userNotOwner) {
```
отступ

20cognosce commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Transactional
+    public void delete(long goalId) {
+        Goal goal = getGoalById(goalId);
```
а в самой цели уже нет её владельца что ли?

20cognosce commented on src/main/java/school/faang/user_service/validator/goal/GoalValidator.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void validateUpdateCompleteGoal(Goal goal) {
+        boolean goalIsComplete = goal.getStatus().equals(GoalStatus.COMPLETED);
```
.

20cognosce commented on src/main/java/school/faang/user_service/validator/goal/GoalValidator.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void validateAllSubGoalsCompleted(long goalId, Stream<Goal> subGoals) {
+        String activeSubGoalIds = subGoals.filter(goal -> !goal.getStatus().equals(GoalStatus.COMPLETED))
```
.

20cognosce commented on src/main/java/school/faang/user_service/validator/goal/GoalValidator.java:
```diff
@@ -0,0 +1,46 @@
+
+@Component
+@RequiredArgsConstructor
+public class GoalValidator {
```
опять у тебя половина проверки в сервиса, а вторая половина в валидаторе

FROLOV-KI committed:
BJS2-77992. Обновлен сервис Целей

FROLOV-KI committed:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-77992-add-goal

FROLOV-KI committed:
BJS2-77992. Обновлен сервис Целей

1) Разделил DTO для создания, обновления и ответа
2) Добавил User сервис
3) Добавил в DTO поле deadline

FROLOV-KI committed:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-77992-add-goal

FROLOV-KI committed:
BJS2-77992. Обновил фильтры

FROLOV-KI committed:
BJS2-77992. Удалил неиспользуемый метод

FROLOV-KI committed:
BJS2-77992. Код стайл

FROLOV-KI committed:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-77992-add-goal

20cognosce commented on src/main/java/school/faang/user_service/entity/goal/dto/request/UpdateGoalDto.java:
```diff
@@ -0,0 +1,19 @@
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public record UpdateGoalDto(
+        @NotBlank(message = "Empty goal title not allowed!") String title,
```
можно не писать подробные сообщения

20cognosce commented on src/main/java/school/faang/user_service/repository/SkillRepository.java:
```diff
@@ -40,6 +40,16 @@ public interface SkillRepository extends JpaRepository<Skill, Long> {
+    @Modifying
+    void assignSkillToGoal(long goalId, long skillId);
+
+    @Query(nativeQuery = true, value = """
+            DELETE FROM goal_skill
+            WHERE goal_id = :goalId""")
```
лучше не нативкой это делать, а через JPA

20cognosce commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -46,5 +46,12 @@ WITH RECURSIVE subtasks AS (
+    List<Long> findUsersByGoalId(long goalId);
+
+    @Modifying
+    @Query(nativeQuery = true, value = """
+            DELETE FROM user_goal ug
+            WHERE ug.user_id = :userId AND ug.goal_id = :goalId
+            """)
```
.

20cognosce commented on src/main/java/school/faang/user_service/validator/goal/SkillValidator.java:
```diff
@@ -0,0 +1,32 @@
+                .toList();
+
+        if (!absentSkillsId.isEmpty()) {
+            String notExistingSkillsId = absentSkillsId.stream()
+                    .map(String::valueOf)
+                    .collect(Collectors.joining(", "));
```
вот это целое действие только ради лога, такое

20cognosce commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,144 @@
+        goalRepository.removeGoalFromUser(userContext.getUserId(), goalId);
+
+        if (goalRepository.findUsersByGoalId(goalId).isEmpty()) {
+            goalRepository.findByParent(goalId).forEach(goalRepository::delete);
```
что со вложенностью-то решили? может сделаешь тогда ограничение, что если у цели есть родитель, то она сама не может быть родителем? чтобы был 1 уровень вложенности

20cognosce reviewed (approved):
Нормально, но чересчур многовато получилось для этой задачи, 700 строк
